infra/gravity/provision.go

```diff
@@ -353,11 +354,34 @@ func bootstrapAzure(ctx context.Context, g *gravity, param cloudDynamicParams) (
 
 // bootstrapCloud is a simple workflow to wait for cloud-init to complete
 func bootstrapCloud(ctx context.Context, g *gravity, param cloudDynamicParams) (err error) {
-	err = sshutil.WaitForFile(ctx, g.Client(), g.Logger(), cloudInitCompleteFile, sshutil.TestRegularFile)
-	if err != nil {
-		return trace.Wrap(err)
-	}
+	return trace.Wrap(g.waitForBootstrapScript(ctx))
+}
 
+func (g *gravity) waitForBootstrapScript(ctx context.Context) error {
+	cmd := fmt.Sprintf(`sudo bash -c "[[ -f %v ]] && exit 0 || [[ -f %v ]] && exit 2 || exit 1"`,
+		cloudInitCompleteFile, cloudInitFailedFile)
+	err := wait.Retry(ctx, func() error {
+		err := sshutil.RunAndParse(ctx, g.Client(), g.Logger(), cmd, nil, sshutil.ParseDiscard)
+		if err == nil {
+			return nil
+		}
+		if exitError, ok := trace.Unwrap(err).(sshutil.ExitStatusError); ok {
+			/*
+			   0      bootstrap successful file found
+			   2      bootstrap failed file found
+			   1      no bootstrap file found
+			*/
+			switch exitError.ExitStatus() {
+			case 0:
+				return nil
+			case 2:
+				return wait.Abort(trace.BadParameter("bootstrap script failed"))
```

Comment:
Here, in the context of explicit retry loop it does not make a difference really which constructor to use as the final word is between `wait.Abort` and `wait.Continue`. This is an API question - `wait.Abort` takes an error as a parameter although it only uses it for the message. I could swap it to a generic `trace.Errorf` or I could make a change to `Abort` API to take a message instead - I think this would be the cleanest although not the smallest of changes.

Reply:
I don't want to delay this change to include Trace work, though it is a reasonable idea.  I think Errorf would be better than BadParameter.